parlai/core/teachers.py

```diff
@@ -1084,3 +1087,84 @@ def setup_data(self, path):
                     reward = 0
             if x:
                 yield [x, None, reward], start
+
+
+class ParlAIDialogTeacher(FixedDialogTeacher):
+    """This module provides access to data in the ParlAI Text Dialog format.
+
+
+    Subclasses ``FixedDialogTeacher`` for functionality and provides an
+    implementation of ``setup_data()`` which iterates over datasets in the
+    "ParlAI text" format. If your data is in the format below, use this class to
+    handle file parsing for you.
+
+    The way the data is set up is as follows:
+
+    ::
+
+        text:Sam went to the kitchen.\nPat gave Sam the milk.\nWhere is the milk?<TAB>labels:kitchen<TAB>reward:1<TAB>label_candidates:hallway|kitchen|bathroom
+        text:Sam went to the hallway.\nPat went to the bathroom.\nWhere is the milk?<TAB>labels:hallway<TAB>reward:1<TAB>label_candidateshallway|kitchen|bathroom<TAB>episode_done:True
+
+    Lines 1-2 represent a single episode, with a different example on each line.
+    The lines contain a query and a label for getting the question
+    correct, and three label candidates.
+
+    Since both of these examples are part of the same episode, the information
+    provided in the first example is relevant to the query in the second
+    example and therefore the agent must remember the first example in order to
+    do well.
+
+    In general dialog this format can contain any speech, not just QA pairs:
+
+    ::
+
+        text:Hi how's it going?<TAB>labels:It's going great. What's new?
+        text:Well I'm working on a new project at work.<TAB>labels:Oh me too!
+        text:Oh cool!<TAB>labels:Tell me about yours.
+
+    etc.
+
+    Note that dialogs are interpreted as being one-way. For example, consider
+    this dialog:
+
+    ::
+
+        1 X1    Y1
+        2 X2    Y2
+        3 X3    Y3
+
+    A set of examples X1 => Y1, X2 => Y2, and X3 => Y3 will be generated.
+    However, Y1 => X2 and Y2 => X3 are not created as separate examples by
+    default. This makes sense for some data (we don't need to train on the idea
+    that "kitchen" should be followed by "Sam went to the hallway..." above),
+    but for other datasets it may be helpful to add additional examples in the
+    reverse direction ("Oh cool!" is a response to "Oh me too!" above).
+    """
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        if shared is None and opt.get('parlaidialogteacher_datafile') is not None:
+            self._setup_data(opt.get('parlaidialogteacher_datafile'))
+        self.id = opt.get('parlaidialogteacher_datafile', 'teacher')
+        self.reset()
+        
+    def num_examples(self):
+        return len(self.examples)
+
+    def num_episodes(self):
+        return self.episodes
+
+    def get(self, episode_idx, entry_idx=None):
```

Comment:
In the case where there are multiple examples per episode (e.g. in your example above), it might be better to keep track of a list of episodes, rather than a list of examples. That is, have self.episodes be a list of episodes, where you can index into the episode with the entry_idx. In this current setup, during random training you may get examples in the middle of an episode, and then jump to a new episode.

Reply:
ah ok, i didnt understand the interface thanks